c2rust-analyze/src/borrowck/type_check.rs

```diff
@@ -33,10 +37,114 @@ impl<'tcx> TypeChecker<'tcx, '_> {
     }
 
     pub fn visit_place(&mut self, pl: Place<'tcx>) -> LTy<'tcx> {
-        let mut lty = self.local_ltys[pl.local.index()];
+        let mut lty: LTy = self.local_ltys[pl.local.index()];
+
+        let mut adt_func = |base_lty: LTy<'tcx>,
+                            base_adt_def: AdtDef,
+                            field: Field,
+                            field_ty: Ty<'tcx>| {
+            let base_origin_param_map: IndexMap<OriginParam, Origin> = base_lty
+                .label
+                .origin_params
+                .map(|params| IndexMap::from_iter(params.to_vec()))
+                .unwrap_or_default();
+
+            let mut field_origin_param_map = IndexMap::new();
+
+            let field_def: &FieldDef = &base_adt_def.non_enum_variant().fields[field.index()];
+            let perm = self.field_permissions[&field_def.did];
+            let base_metadata = &self.adt_metadata.table[&base_adt_def.did()];
+            let field_metadata = &base_metadata.field_info[&field_def.did];
+
+            if let Some(field_adt_metadata) = self.adt_metadata.table.get(&field_def.did) {
+                /*
+                    If we're in this block, it means the current field is an ADT.
+
+                    That field's type may have its own lifetime parameters. In the following:
+
+                    ```
+                    struct Foo<'a> {
+                        a: &'a i32
+                    }
+
+                    struct Bar<'b> {
+                        foo: Foo<'b>
+                    }
+
+                    fn some_func() {
+                        let bar: Bar<'0> = ...;
+                        bar.foo.a: &'? i32 = ...;
+                    }
+                    ```
+
+                    We want to know that the lifetime `'?` gets resolved to the concrete
+                    origin `'0`. To do this, a mapping needs to be made between `bar.foo`
+                    lifetime parameter `'b` (which is already paired with concrete lifetime
+                    `'0`) and `Foo` lifetime parameter `'a`. This mapping is created below.
+                */
+                for (field_lifetime_arg, field_struct_lifetime_param) in field_metadata
+                    .lifetime_params
+                    .iter()
+                    .zip(field_adt_metadata.lifetime_params.iter())
+                {
+                    let field_lifetime_param =
+                        if let Ok(param) = OriginParam::try_from(field_lifetime_arg) {
+                            param
+                        } else {
+                            continue;
+                        };
```

Comment:
I think what we're doing in this loop is essentially evaluating `field_lifetime_arg`, which is a sort of "origin expression", down to a concrete `Origin` that we can store in `field_origin_param_map` under the key `field_struct_lifetime_param`.  The evaluator has to handle all three kinds of "expressions": `ReEarlyBound` (resolved by looking in `base_origin_param_map`), `Hypothetical` (also resolved by looking in `base_origin_param_map`), and `ReStatic` (currently ignored - the `continue` here).  I'm not sure how `'static` is represented in Polonius, but we should either do it properly or complain that it's unsupported.

Reply:
It's represented as a universal region. I've put in panics  for now in 6465c22c, and would like to add support for this in a follow-on PR.

Reply:
Sounds good.